### PR TITLE
wip sign command

### LINKED ISSUE #5: Sign and verify  something
Add signing and verification of data through public key.

### PR DESCRIPTION
**Pull request by marti1125**
it is a WIP

I just called  this method

 Sign(privateKey *[32]byte, message []byte, random [64]byte)

but I don't have idea how to create [32]byte and random [64]byte 
from []byte.

please check my initial version

### CODE REVIEW

**Comment by sah4ez** [comment]
Oh, it's very simple.
For `[64]byte random` you can use something like this (PRNG):
```go
func random() [64]byte {
    var random [64]byte
    r := rand.Reader
    io.ReadFull(r, random[:])
    return random
}
```
For work with array bytes you can write and use this utility function:
```go
func slice2array32(src []byte) [32]byte {
    var result [32]byte
    copy(result[:], src[:32])
    return result
}
```
Also, you can write generic functions for other often usefull power of 2 (8, 16, 32, 64 and 128).
As example:
```go
priv, pub := GenerateDH()
message := []byte("message for signature")
privArray := slice2array32(priv)
signature := Sign(&privArray, message, random())
```
Will be cool, if you write unit tests for this function. And remember, that need user `copy` all inputs slice bytes for signature and verification.

**Comment by marti1125** [comment]
I am using this part of code for sign a message, like publish command

pub, priv, err := keys.GenerateDH()
sign := keys.Sign(pub, []byte(strings.Join(message, " ")), utils.Random())

but i verify command always return false, I printed in console the keys from  keys.GenerateDH()
is very different from https://freecontent.space/keys.html
search find by name: will


````
pub, err := api.Load(keyName)
			if err != nil {
				return errors.Wrap(err, "can not load public name")
			}

			pubArray := utils.Slice2Array32(pub)
			random := utils.Random()

			t := base64.StdEncoding.EncodeToString(pubArray[:])
			fmt.Println(t)

			verify := keys.Verify(pubArray, []byte(strings.Join(message, " ")), &random)
			fmt.Fprintln(out, verify)
			return nil
````

what is the workflow =?

**Comment by marti1125** [comment]
```
(base) willy.aguirre@localhost:~/go/src/github.com/marti1125/pspk/bin$ ./pspk --name hg sign wally
BDJLNTU4Y6PN3RXKTKO36F32565OEPHBGLONY245ZDFJLRTL5NDQ====
H2UJLRRSFR7L7JVKRDDXKWDJBTPMDLQU64HNOLQPSYXWV6BK5EKQ====
QDD3saWLmFUdgU78Urp/zvnYzkzTo8iKUH1qu0uB2c+A/brop7gAqdOPhtZdZkGgSgTIDw6I3x4dOKZniN0fDg==

(base) willy.aguirre@localhost:~/go/src/github.com/marti1125/pspk/bin$ ./pspk verify hg wally
PqiVxjIsfr+mqojHdVhpDN7BrhT3DtcuD5Yvavgq6RU=
CNK2zpzHnt3G6pqdvxd677riPOEy3NxrncjKlcZr60c=
RqySwoiRY55nb0IBs7puCE7+79v9n/lLqBb9pACfpEw=
bawmyruoPoGG6SNQ3syf77zfT8FhNhYdmbiDTnGWgLQrgnkPBcueVQ3j2+8y3unf3czpryjVItoZYqkZKKmcLPdpSVjkuaBbpBftEcZiWNo=
false

```

**Comment by sah4ez** [comment]
1) All signatures need make  with private key
2) Verification of signatures need make with public key of signatory
So, first.
If you don't have a keys on server, first need publish key to the server:
```
pspk --name <YOU_NAME> publish
```
The next, you can make sign data with private key, as example:
```
pspk --name <YOU_NAME> sign <MESSAGE_WITH_SPACES>
```
by `<YOU_NAME>` pspk would found private key and made signature. The output of this command will be a signature in base64.
The next you should put this signature with message or file to the verification command, as example:
```
pspk verify <YOU_NAME> <SIGNATURE_IN_BASE64> <MESSAGE_WITH_SPACES>
```
by <YOU_NAME> pspk found public key on the server, download it and made verification of the signature of the message. And need print some text about result of operation, as example:
```
pspk veriyf gh base64_of_signature <SIGNATURE_IN_BASE64> <MESSAGE_WITH_SPACES>
Signature <SIGNATURE_IN_BASE64> is valid.
or
Signature <SIGNATURE_IN_BASE64> is NOT valid.
```

File: cmd/cli/sign.go [hunk 1]
```diff
@@ -24,12 +28,32 @@ func Sign() cli.Command {
...
-			priv, err := utils.Read(path, "key.bin")
+			pub, priv, err := keys.GenerateDH()
+			if err != nil {
+				return errors.Wrap(err, "can not generate keys")
+			}
+			err = utils.Write(path, "pub.bin", pub[:])
+			if err != nil {
+				return errors.Wrap(err, "can not write in pub.bin")
+			}
+			err = api.Publish(name, pub[:])
```
Comment (sah4ez) [inline cmd/cli/sign.go]: Nope, you didn't need publish any keys in sign command. `publish` ans `sign` - it's different commands, where output of `publish` can be used in the `sign` command

File: cmd/cli/sign.go [hunk 2]
```diff
@@ -24,12 +28,32 @@ func Sign() cli.Command {
...
+			err = utils.Write(path, "key.bin", priv[:])
+			if err != nil {
+				return errors.Wrap(err, "can not find the path")
+			}
 
 			if err != nil {
 				return errors.Wrap(err, "can not read key.bin")
 			}
 
+			sign := keys.Sign(pub, []byte(strings.Join(message, " ")), utils.Random())
```
Comment (sah4ez) [inline cmd/cli/sign.go]: Use a private key for signatures.

File: cmd/cli/verify.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		Name:        "verify",
+		Aliases:     []string{"v"},
+		Description: "Verify the message through ed25519",
+		Usage:       "verify <KEY_NAME> <MESSAGE>",
+		Action: func(c *cli.Context) error {
+			keyName := c.Args()[0]
+			message := c.Args()[1:]
+
+
+			path = path + "/" + keyName
```
Comment (sah4ez) [inline cmd/cli/verify.go]: In `verify` you didn't load any keys from you localhost machine.

**Review by sah4ez — CHANGES_REQUESTED** [review]
Rule of signatures:
sign with private key and verify with public key.

**Comment by marti1125** [comment]
Thanks for the workflow :+1: 

my progress  is:


```
(base) willy.aguirre@localhost:~/go/src/github.com/marti1125/pspk/bin$ ./pspk --name willy publish
Generate key pair on x25519

(base) willy.aguirre@localhost:~/go/src/github.com/marti1125/pspk/bin$ ./pspk --name willy sign hello world!

4riJPiF7vDr5BgP1fqXB6nJQ2FfXeBL/0r3R2Fd2RZYn8g1jdOUpKsFYcOQSsylUH9XjbRWmZgeKohbIy1Y0Aw==

(base) willy.aguirre@localhost:~/go/src/github.com/marti1125/pspk/bin$ ./pspk verify willy 4riJPiF7vDr5BgP1fqXB6nJQ2FfXeBL/0r3R2Fd2RZYn8g1jdOUpKsFYcOQSsylUH9XjbRWmZgeKohbIy1Y0Aw== hello world!

Signature 4riJPiF7vDr5BgP1fqXB6nJQ2FfXeBL/0r3R2Fd2RZYn8g1jdOUpKsFYcOQSsylUH9XjbRWmZgeKohbIy1Y0Aw== is valid.
```

File: cmd/cli/sign.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+			path = path + "/" + name
+
+			priv, err := utils.Read(path, "key.bin")
+			if err != nil {
+				return errors.Wrap(err, "can not read key.bin")
+			}
+
+			privArray := utils.Slice2Array32(priv)
+
+			sign := keys.Sign(&privArray, []byte(strings.Join(message, " ")), utils.Random())
```
Comment (sah4ez) [inline cmd/cli/sign.go:38]: You used `keys.Sign` but I didn't see this changes in PR. :-)

**Review by sah4ez — CHANGES_REQUESTED** [review]

File: cmd/cli/sign.go (hunk 4 above)
Comment (marti1125) [reply cmd/cli/sign.go:38]: I changed that part I am using the private key for signature and key.Sign
```
priv, err := utils.Read(path, "key.bin")
			if err != nil {
				return errors.Wrap(err, "can not read key.bin")
			}

			privArray := utils.Slice2Array32(priv)

			sign := keys.Sign(&privArray, []byte(strings.Join(message, " ")), utils.Random())
			data := base64.StdEncoding.EncodeToString(sign[:])
```

Is it correct ?

Comment (sah4ez) [reply cmd/cli/sign.go:38]: Yes, it's correct. 
But I didn't see changes in the `key` package, there should be `Sign` and `verify` methods.

Comment (marti1125) [reply cmd/cli/sign.go:38]: I am using from https://github.com/sah4ez/pspk/blob/06fbc8d0d2d2f2925e30bdc9afe731701c0e3fb8/pkg/keys/keys.go

Comment (marti1125) [reply cmd/cli/sign.go:38]: could you try it my branch?

Comment (sah4ez) [reply cmd/cli/sign.go:38]: Oh... I forgot that I added it xD
That all ok, I'll merge it today. Thanks!

Comment (sah4ez) [reply cmd/cli/sign.go:38]: Oh... I forgot that I added it xD
That all ok, I'll merge it today. Thanks!

Comment (marti1125) [reply cmd/cli/sign.go:38]: I forgot the part of --file in the command `sign --file <path_to_file>`
could be in another PR?
and could you give an example about  `sign --file <path_to_file>`
Thanks!

Comment (sah4ez) [reply cmd/cli/sign.go:38]: Yes, you can create another PR with support `--file` flag.

**Review by sah4ez — APPROVED** [review]

**Pull request merged**